### PR TITLE
vae examples and benchmarks vs pytorch (#1289)

### PR DESCRIPTION
**Pull request by pw2393**
A vae example and benchmark vs pytorch

### CODE REVIEW

**Comment by pmichel31415** [comment]
That's very cool! From what I understand dynet is slower than pytorch on CPU with macOS, do you have any intuition why? maybe related to the Eigen version or something?

**Comment by xunzhang** [comment]
Eigen version is a possibility. Another possibility is compiling options.

**Comment by neubig** [comment]
Thanks, this is good! I've merged this because I think it'd be useful now as-is.

1. It seems that the loss is slightly worse in all reported results. Do you think this is a real difference, i.e. is it consistent across multiple runs? If so, we should probably figure out why. It might be due to parameter initialization.
2. We can open another issue to examine why DyNet is slower on the Mac.

**Comment by pw2393** [comment]
Thanks!

1. Excuse me about the confusing 4-decimal test loss reported. It should have been `97.3±0.X` (of which the randomness comes from the parameter initialization and the prior), for both DyNet and PyTorch after 10 epochs of training. (During multiple runs, one's test loss could be lower than the other one's, while sometimes in reverse. The DyNet final reconstructed-or-sampled images are as good as those of PyTorch. )

2. The future benchmarks could also possibly offer some hints for it (the mac slow-down).

**Comment by pw2393** [comment]
**UPDATE: ABOUT THE MAC OS**

The original CPU benchmarks on MAC OS were based on the current PyPI DyNet package. The environment was built via `pip install dynet`.

However, if I build DyNet manually with 
> https://bitbucket.org/eigen/eigen at `2355b22`
> https://github.com/clab/dynet.git at master d69bdc1

the results now become (the bolded):

| OS | Device | Framework | Speed |
| --- | --- | --- | --- |
| MAC OS 10.13.3 | 2.3 GHz Intel Core i5 | PyTorch | ~ 12.5±0.3 s per epoch |
| MAC OS 10.13.3 | 2.3 GHz Intel Core i5 | DyNet (PyPI) | ~ 19.8±0.3 s per epoch |
| MAC OS 10.13.3 | 2.3 GHz Intel Core i5 | DyNet (**Master**) | ~ **9.9±0.3 s per epoch** | 

On Ubuntu, no matter how I install DyNet (via pypi or github master), the results remain the same.

**Comment by zhisbug** [comment]
@pw2393 How did you install Pytorch? This is probably not a fair claim if you install pytorch directly using their binaries.

**Comment by pw2393** [comment]
screen shot 2018-03-26 at 3 53 09 pm

**Comment by zhisbug** [comment]
Is it possible to compile PyTorch from source and see how it performs on Mac?
